Support connection `Upgrade` and `CONNECT`.

MtkN1 opened the pull request:
<!-- Thanks for contributing to HTTP Core! 💚
Given this is a project maintained by volunteers, please read this template to not waste your time, or ours! 😁 -->

# Summary

<!-- Write a small summary about what is happening here. -->

- Fixes #872

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [ ] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

MtkN1 commented:
<!-- 作業を開始する為にドラフトプルリクエストを作成しました！
まず `Upgrade`/`CONNECT` で例外を発生するようにしています。

私の理解では次の作業が必要です ...

1. こちらのアドバイス (https://github.com/encode/httpcore/discussions/871#discussioncomment-8223419) のような trailing data を処理できるネットワークストリームを実装する
2. 例外を発生させた箇所でそのネットワークストリームを生成する
3. `network_stream` に格納する変数を切り替える
    https://github.com/encode/httpcore/blob/79fa6bf0dfcf3820d1ae7e52a2d268f33022c5a4/httpcore/_async/http11.py#L125

何かアドバイスを頂けると幸いです！ -->

I have created a draft pull request to begin the work!
First, `Upgrade`/`CONNECT` (incorrectly) raises an exception.

My understanding is that the following work needs to be done...

1. Implement a network stream that can handle trailing data like this advice (https://github.com/encode/httpcore/discussions/871#discussioncomment-8223419)
2. Generates its network stream at the step that raised the exception
4. Switch variables to be stored in `network_stream`.
    https://github.com/encode/httpcore/blob/79fa6bf0dfcf3820d1ae7e52a2d268f33022c5a4/httpcore/_async/http11.py#L125

@tomchristie  I would appreciate any advice you can give me!

tomchristie commented:
Ah fantastic thanks, shall we start with a test?...

1. There's an existing `Upgrade` test here, that could be replicated into a new `test_upgrade_with_trailing_data` test case.

https://github.com/encode/httpcore/blob/79fa6bf0dfcf3820d1ae7e52a2d268f33022c5a4/tests/_async/test_http11.py#L238
 
I think the next step after that will be...

2. We probably want to change `receive_response_headers` so that it also returns the `trailing_data` as part of the response info, include the trailing data in the raised exception.

And then...

3. We'll want an `HTTPUpgradeNetworkStream(SyncNetworkStream)` class. This have an `__init__` method taking a network stream and trailing data. Most of the methods should delegate to the underlying network stream, except `read` which should initially return the trailing data. We'll return that in the extensions for the `Upgrade`/`CONNECT` cases.

MtkN1 commented:
Thank you for your guidance 🙏 I'll give it a try.

MtkN1 commented:
<!--
作業が完了したので私は準備完了にマークしました。 レビューお願いします！

- `HTTP11UpgradeStream` を追加しました。
  - `read()` メソッドで trailing data を受信することができます。
  - 他のメソッドは元の NetworkStream にバイパスします。
- `Upgrade` と `CONNECT` では `network_stream` 拡張機能を `HTTP11UpgradeStream` に切り替えます。
- テストを追加しました。 100% カバーしています。
- この修正によって私のディスカッション (https://github.com/encode/httpcore/discussions/871#discussion-6123741) と httpx-ws の問題 (https://github.com/frankie567/httpx-ws/issues/7#issue-1463219741) が解決されることを確認しました。
- この変更はユーザーが期待する動作になるので、私はドキュメントを追加していません。
-->

Work completed, I marked PR as ready. Could you please review this?

- Added `HTTP11UpgradeStream`.
  - The `read()` method allows trailing data to be received.
  - Other methods bypass the original NetworkStream.
- For `Upgrade` and `CONNECT`, switch the `network_stream` extension to `HTTP11UpgradeStream`.
- Added tests. 100% coverage.
- I confirm that this fix resolves my discussion (https://github.com/encode/httpcore/discussions/871#discussion-6123741) and httpx-ws issue (https://github.com/frankie567/httpx-ws/issues/7#issue-1463219741).
- I am not adding documentation because this change would be the behavior the user expects.

tomchristie commented on tests/_async/test_http11.py:
```diff
@@ -265,10 +265,65 @@ async def test_http11_upgrade_connection():
+            assert response.status == 101
+            network_stream = response.extensions["network_stream"]
+
+            assert network_stream._leading_data != b""
+            content = await network_stream.read(max_bytes=3)
+            assert content == b"foo"
+            content = await network_stream.read(max_bytes=3)
+            assert content == b"bar"
+
+            assert network_stream._leading_data == b""
+            content = await network_stream.read(max_bytes=3)
+            assert content == b"baz"
```
Test is great.

Let's not assert on the private state here, the read/content is enough.

MtkN1 replied:
Understood. I have modified it as suggested!

tomchristie commented on tests/_async/test_http11.py:
```diff
@@ -265,10 +265,65 @@ async def test_http11_upgrade_connection():
         ) as response:
             assert response.status == 101
             network_stream = response.extensions["network_stream"]
+            assert network_stream._leading_data == b""
```
```suggestion
```

tomchristie reviewed (commented):
This one is a fantastic fix, thanks. ✨
Minor suggestion wrt. the tests.

---

Later follow-ups we can consider...

* Should all HTTP requests include this behaviour or only `Upgrade`/`CONNECT`?
* Should all HTTP requests return `network_stream` or only `Upgrade`/`CONNECT`?
* What might a nice `__repr__` be here?

(I'm just jotting these down here for reference)

T-256 commented on httpcore/_async/http11.py:
```diff
@@ -340,3 +353,34 @@ async def aclose(self) -> None:
                 await self._connection._response_closed()
+
+
+class HTTP11UpgradeStream(AsyncNetworkStream):
```
- is this only for `Upgrade`?
- is this only for http1.1?
- how about adding `Async` prefix
- Perhaps could move it to `httpcore._backends` into `sync` and `auto` submodules.

T-256 commented on httpcore/_async/http11.py:
```diff
@@ -340,3 +353,34 @@ async def aclose(self) -> None:
                 await self._connection._response_closed()
+
+
+class HTTP11UpgradeStream(AsyncNetworkStream):
```
```suggestion
class AsyncHTTP11TrailingStream(AsyncNetworkStream):
```

MtkN1 replied:
<!--
Async の接頭辞を付けるべきなのでしょうか？  同じ `http11.py` ファイルにある `HTTP11ConnectionByteStream` クラスには接頭辞がありません。
他のファイルのユーティリティクラスでも同様のパターンがあります。
-->

Should I prefix it with Async?  The `HTTP11ConnectionByteStream` class in the same `http11.py` file has no prefix.
A similar pattern can be seen in the utility classes of other files.

tomchristie replied:
Yep. The `scripts/unasync` script will strip that portion in the corresponding `_sync` code.

MtkN1 replied:
Added. `scripts/unasync` is great all around!

tomchristie commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 
+## Unreleased
+
+- Support connection Upgrade and CONNECT. (#882)
```
```suggestion
- Fix support for connection Upgrade and CONNECT when some data in the stream has been read. (#882)
```

T-256 commented on tests/_async/test_http11.py:
```diff
@@ -269,6 +269,57 @@ async def test_http11_upgrade_connection():
+            content = await network_stream.read(max_bytes=3)
+            assert content == b"baz"
+
+            # Lazy tests for HTTP11UpgradeStream
```
```suggestion
            # Lazy tests for AsyncHTTP11UpgradeStream
```

MtkN1 replied:
Oops, thank you!

T-256 reviewed (approved):
LGTM, Thank You!

MtkN1 commented:
Thank you for your approval 🚀 It is a pleasure to contribute to this project.
